astropy/coordinates/baseframe.py

```diff
@@ -1581,6 +1581,36 @@ def __setattr__(self, attr, value):
 
         super().__setattr__(attr, value)
 
+    def __eq__(self, value):
+        """Equality operator for frame.
+
+        This implements strict equality and requires that the frames are
+        equivalent and that the representation data are exactly equal.
+        """
+        if not self.is_equivalent_frame(value):
+            return False
+
+        if value._data is None and self._data is not None:
+            return False
+
+        if self._data is None and value._data is not None:
+            return False
+
+        if self._data is None and value._data is None:
+            return True
+
+        if self.shape != value.shape:
+            return False
+
+        return self._data == value._data
+
+    def __ne__(self, value):
+        eq = self.__eq__(value)
+        if isinstance(eq, bool):
+            return not eq
+        else:
+            return ~eq
```

Comment:
Wow, never appreciated that `~True` equals `-2`!

Reply:
That surprised me too!

Reply:
what? how? why? (I could see it it were -1, but -2?)

Reply:
As @jdavies-st explained it to me, `~` flips the bits and somehow `True` is treated as `int` when `~` is applied.
```python
>>> np.binary_repr(1, width=8)                                                                                                      
'00000001'
>>> np.binary_repr(-2, width=8)                                                                                                     
'11111110'
```

Reply:
Isn't `__ne__` automagically defined if `__eq__` is defined in python3?  Does `__eq__` as defined above ever return something other than `True` or `False`?

Btw, I'm really happy to see `__eq__` for strict equality implemented for `SkyCoord`. Will make testing much cleaner. 🎉

Reply:
I'm not sure the automagic `__ne__` would be numpy-aware, which is necessary since it can return a numpy array of bool type.  Also, SkyCoord already has a `__ne__` method that comes from *somewhere* but I honestly don't know where.
```
In [7]: sc.__ne__??                                                                                                                          
Signature:      sc.__ne__(value, /)
Call signature: sc.__ne__(*args, **kwargs)
Type:           method-wrapper
String form:    <method-wrapper '__ne__' of SkyCoord object at 0x7f7f29204850>
Docstring:      Return self!=value.
```

Reply:
I see.  That makes sense.

```
In [13]: ~np.array([True, True, False])
Out[13]: array([False, False,  True])
```

Thanks.  I'd forgot that `~` acting on numpy boolean arrays is the same as `np.logical_not`. 👍

Reply:
Thanks for the reminder on `np.logical_not`.  That works on both scalar and np.array values so I got rid of that ugly code.